src/builder.rs

```diff
@@ -664,66 +717,86 @@ pub fn bundle<V: TryFrom<i64>>(
-        .map_err(|_| BuildError::BundleTypeNotSatisfiable)?;
+    // The number of actions to add to this bundle in order to contain at least MIN_ACTION actions.
+    let num_missing_actions = MIN_ACTIONS.saturating_sub(spends.len().max(outputs.len()));
 
     // Pair up the spends and outputs, extending with dummy values as necessary.
-    let (pre_actions, bundle_meta) = {
-        let mut indexed_spends = spends
-            .into_iter()
-            .chain(iter::repeat_with(|| SpendInfo::dummy(&mut rng)))
-            .enumerate()
-            .take(num_actions)
-            .collect::<Vec<_>>();
-
-        let mut indexed_outputs = outputs
-            .into_iter()
-            .chain(iter::repeat_with(|| OutputInfo::dummy(&mut rng)))
-            .enumerate()
-            .take(num_actions)
-            .collect::<Vec<_>>();
-
-        // Shuffle the spends and outputs, so that learning the position of a
-        // specific spent note or output note doesn't reveal anything on its own about
-        // the meaning of that note in the transaction context.
-        indexed_spends.shuffle(&mut rng);
-        indexed_outputs.shuffle(&mut rng);
-
-        let mut bundle_meta = BundleMetadata::new(num_requested_spends, num_requested_outputs);
-        let pre_actions = indexed_spends
-            .into_iter()
-            .zip(indexed_outputs.into_iter())
-            .enumerate()
-            .map(|(action_idx, ((spend_idx, spend), (out_idx, output)))| {
-                // Record the post-randomization spend location
-                if spend_idx < num_requested_spends {
-                    bundle_meta.spend_indices[spend_idx] = action_idx;
-                }
-
-                // Record the post-randomization output location
-                if out_idx < num_requested_outputs {
-                    bundle_meta.output_indices[out_idx] = action_idx;
-                }
-
-                ActionInfo::new(spend, output, &mut rng)
-            })
-            .collect::<Vec<_>>();
-
-        (pre_actions, bundle_meta)
-    };
+    let (pre_actions, bundle_meta, _) = partition_by_asset(&spends, &outputs, &mut rng)
+        .into_iter()
+        .fold(
+            (
+                Vec::new(),
+                BundleMetadata::new(num_requested_spends, num_requested_outputs),
+                // We might have to add dummy/split actions only for the first asset to reach MIN_ACTIONS.
+                num_missing_actions,
+            ),
+            |(mut pre_actions, mut bundle_meta, num_missing_actions),
+             (asset, (spends, outputs))| {
+                let num_actions = spends.len().max(outputs.len()) + num_missing_actions;
+
+                let first_spend = spends.first().map(|(_, s)| s.clone());
+
+                let mut indexed_spends = spends
+                    .into_iter()
+                    .chain(iter::repeat_with(|| {
+                        (usize::MAX, pad_spend(first_spend.as_ref(), asset, &mut rng))
```

Comment:
`usize::MAX` is unintuitive, use named var / constant (everywhere)
 reading the code I need to understand **at this point** that this is the spend index but currently it is unclear
also unclear how multiple spends can have the same index `(usize:max)`